omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/ConceptResource1_8.java

```diff
@@ -256,59 +262,81 @@ public static void setFullySpecifiedName(Concept instance, String name) {
+		String removeMethodName ="remove"+elementName;
+		
+		Method addMethod = ReflectionUtil.findMethod(Concept.class, addMethodName);
+		Method removeMethod = ReflectionUtil.findMethod(Concept.class, removeMethodName);
+		
+		for(T value : items){
+			PropertyUtils.setProperty(value, "concept", instance);
 		}
-		//delete names which are absent new list of names
-		for (ConceptName name : instance.getNames()){
-			if(!checkIfNamesContainNameByProperties(names, name)){
-				instance.removeName(name);
+		//delete object which are absent in new list
+		removeRedundant:
+		for(T oldValue : (Collection<T>)PropertyUtils.getProperty(instance, propertyName)){
+			for(T newValue : items){
+				if(itemsComparator.compare(oldValue, newValue)==0){
+					continue removeRedundant;
+				}
 			}
+			removeMethod.invoke(instance, oldValue);
 		}
-		//add names which are absent in old list of names
-		for(ConceptName name : names){
-			if(!checkIfNamesContainNameByProperties(instance.getNames(), name)){
-				instance.addName(name);
+		//add object which are absent in old list
+		addNew:
+		for(T newValue : items){
+			for(T oldValue : (Collection<T>)PropertyUtils.getProperty(instance, propertyName)){
+				if(itemsComparator.compare(oldValue, newValue)==0){
+					continue addNew;
+				}
 			}
+			addMethod.invoke(instance, newValue);
 		}
 	}
+	
 	/**
-	 * helper method for {@link #setNames(Concept, List<ConceptName>)}
-	 * checks if specified name is present in concept's names list
+	 * It's needed, because of ConversionException: Don't know how to handle collection class:
+	 * interface java.util.Collection
 	 * 
-	 * @should return true if list contains name with given uuid
-	 * @should return true if list contains name with same properties
-	 * @should return false if there is no concept with same properties
+	 * if request to update Concept updates ConceptName, adequate resource takes care of it,
+	 * so this method just adds new and removes deleted names.
 	 * 
+	 * @param instance
 	 * @param names
-	 * @param searched
-	 * @return
+	 * @throws NoSuchMethodException 
+	 * @throws InvocationTargetException 
+	 * @throws IllegalAccessException 
 	 */
-	public static boolean checkIfNamesContainNameByProperties(Collection<ConceptName> names, ConceptName searched) {
-		for(ConceptName name : names){
-			if(Objects.equals(name.getUuid(), searched.getUuid())){
-				return true;
-			}
-			else if(Objects.equals(name.getName(), searched.getName())
-					&&Objects.equals(name.getConceptNameType(), searched.getConceptNameType())
-					&&Objects.equals(name.getLocale(), searched.getLocale())){
-				return true;
+	@PropertySetter("names")
+	public static void setNames(Concept instance, List<ConceptName> names) 
+			throws IllegalAccessException, InvocationTargetException, NoSuchMethodException {
+    	setHibernateCollectionProperty(instance, "names", names, new Comparator<ConceptName>(){
+			@Override
+			public int compare(ConceptName left, ConceptName right) {

```

Comment:
I wouldn't suggest that to avoid obscure bugs since uuid based equality is widely used all over the place, alternatively for elegance we can add a ConceptName comparator that uses these other semantics but that's up to him if they got time, in any case this is only used here so it might not be worth the effort